Source/NSSwitch.m

```diff
@@ -0,0 +1,288 @@
+    }
+}
+
+- (double) doubleValue
+{
+  return (double)(([self state] == NSControlStateValueOn) ? 1.0 : 0.0);
+}
+
+- (void) setFloatValue: (float)val
+{
+  [self setDoubleValue: (double)val];
+}
+
+- (float) floatValue
+{
+  return (float)[self doubleValue];
+}
+
+- (void) setIntValue: (int)val
+{
+  [self setDoubleValue: (double)val];
+}
+
+- (int) intValue
+{
+  return (int)[self doubleValue];
+}
+
+- (void) setIntegerValue: (NSInteger)val
+{
+  [self setDoubleValue: (double)val];
+}
+
+- (NSInteger) integerValue
+{
+  return (NSInteger)[self doubleValue];
+}
+
+- (void) setStringValue: (NSString *)val
+{
+  [self setDoubleValue: [val doubleValue]];
+}
+
+- (NSString *) stringValue
+{
+  return [NSString stringWithFormat: @"%ld", [self integerValue]];
+}
+
+- (void) setObjectValue: (id)obj
+{
+  if ([obj respondsToSelector: @selector(stringValue)])
+    {
+      [self setStringValue: [obj stringValue]];
+    }
+}
+
+- (id) objectValue
+{
+  return [self stringValue];
+}
+
+- (void) drawRect: (NSRect)rect
+{
+  [[GSTheme theme] drawSwitchInRect: rect
+                           forState: _state
+                            enabled: [self isEnabled]];
+}
+
+- (void) mouseDown: (NSEvent *)event
+{
+  if (![self isEnabled])
+    {
+      [super mouseDown: event];
+      return;
+    }
+  
+  if (_state == NSControlStateValueOn)
+    {
+      [self setState: NSControlStateValueOff];
+    }
+  else
+    {
+      [self setState: NSControlStateValueOn];
+    }
+
+  if (_action)
+    {
+      [self sendAction: _action
+                    to: _target];
+    }
+}
+
+// Accessibility
+- (NSRect)accessibilityFrame
+{
+  return [self frame];
+}
+
+- (NSString *)accessibilityIdentifier
+{
+  return nil;
+}
+
+- (id)accessibilityParent
+{
+  return nil;
+}
+
+- (BOOL)isAccessibilityFocused
+{
+  return NO;
+}
+
+- (NSString *)accessibilityLabel
+{
+  return nil;
+}
+
+- (BOOL)accessibilityPerformPress
+{
+  return NO;
+}
+
+- (BOOL) accessibilityPerformDecrement
+{
+  return NO;
+}
+
+- (BOOL) accessibilityPerformIncrement
+{
+  return NO;
+}
+
+- (NSString *) accessibilityValue
+{
+  return nil;
+}
+
+// NSCoding
+- (id) initWithCoder: (NSCoder *)coder
+{
+  if ((self = [super initWithCoder: coder]) != nil)
+    {
+      if ([coder allowsKeyedCoding])
+        {
+          if ([coder containsValueForKey: @"NSControlContents"])
+            {
+              [self setState: [coder decodeIntegerForKey: @"NSControlContents"]];
+            }
+          if ([coder containsValueForKey: @"NSControlAction"])
+            {
+              NSString *s = [coder decodeObjectForKey: @"NSControlAction"];
+              [self setAction: NSSelectorFromString(s)];
+            }
+          if ([coder containsValueForKey: @"NSControlTarget"])
+            {
+              id t = [coder decodeObjectForKey: @"NSControlTarget"];
+              [self setTarget: t];
+            }
+         if ([coder containsValueForKey: @"NSEnabled"])
```

Comment:
I think NSControl already decodes that, we don't need it here.
The same is true for the encoding.

Reply:
See the comment I left.  The value is decoded in NSControl, but not used.